### DISCUSSION TITLE
Dynamic import of resources/assets with variable path is broken because such imports are always processed as not having a file extension and are being treated as modules

### DISCUSSION THREAD

**Discussion by Sap333** [webpack/webpack#16874] [Q&A]
# Bug report

**What is the current behavior?**

It's impossible to perform a dynamic import of a non-JS module (a PNG picture for example) using variable paths.<br/>
The next code will not add a file to a bundle, even if ithe file exists:
>     const name="picture";
>     import(  `./resources/${name}.png`);

The problem is in the current module resolving algorithm. I quote the [documentation](https://webpack.js.org/concepts/module-resolution/#module-paths):

> Once the path is resolved based on the above rule, the resolver checks to see if the path points to a file or a directory. If the path points to a file:
> 
> If the path has a file extension, then the file is bundled straightaway.
> Otherwise, the file extension is resolved using the [resolve.extensions](https://webpack.js.org/configuration/resolve/#resolveextensions) option, which tells the resolver which extensions are acceptable for resolution e.g. .js, .jsx.

But if an import has a variable path (no matter which kind - even if it's a full path with an insertion of a variable containing an empty string), webpack decides that there is no extension and performs a module resolution, which means that the file must have a valid module extension. So only files with ".js", ".ts" and are included into a bundle. The module resolution performed before any module rules are checked but after include/exclude import rules.

**What is the expected behavior?**

The documentation states and the commons sense tells it must be possible to import other module types than plain JS-modules. If an extension is being provided in the import path, then it must be taken in account. And a specified extension it seems to be the most popular and common case when a non-module is dynamically imported. If no extension is provided/detected then the imported files must be checked and if there are any non-empty "module"-extensions then the module resolution must be performed. The good question is: What should be done if both JS-modules and other file types are detected? There are two main possible choices: not import non-modules in a module has been found (but it can be tricky to execute if a whole subtree is imported), or just import everything and let the user decide (but I suppose that there can be some side effects with importing complex modules). Because of that, I see as the most simple solution is when, webpack might require to always specify an extension if anything else apart from a module is being imported with dynamic import AND variable path. If a user wants several file types to be imported, they can use multiple import statements or use other ways to force webpack to include the files/

**Other relevant information:**
webpack version:  5.76.3
Node.js version: 18.15.0
Operating System:  Linux
Additional tools:

**Comment by alexander-akait** [comment]
Can you provide reproducible test repo, because we have a lot of tests on such scenario and it works correctly
